### DISCUSSION TITLE
How to use OSC52 Paste inside tmux?

### DISCUSSION THREAD

**Discussion by juru1234** [neovim/neovim#29350] [Q&A]
```
vim.g.clipboard = {
  name = 'OSC 52',
  copy = {
    ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
    ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
  },
  paste = {
    ['+'] = require('vim.ui.clipboard.osc52').paste('+'),
    ['*'] = require('vim.ui.clipboard.osc52').paste('*'),
  },
}
```

Hi, i really like the new default OSC52 provider in neovim. Copy and Paste works for me with foot terminal. But I do not get the Paste to work inside a tmux session. Neovim prints the following when pasting:

>  Waiting for OSC 52 response from the terminal. Press Ctrl-C to interrupt...

Any ideas what to set in the tmux config to make this work?

I already tried : `set-window-option -g allow-passthrough on`

**Comment by andreypopp** [comment]
I suggest to use tmux clipboard provider in neovim if tmux is active, but you'd need to call `tmux refresh-client -l` before pasting so tmux queries terminal using OSC52 beforehand. Try a config like this (assuming you have a recent tmux with OSC52 support):
```
vim.g.clipboard = {
  name = 'OSC 52',
  copy = {
    ['+'] = require('vim.ui.clipboard.osc52').copy('+'),
    ['*'] = require('vim.ui.clipboard.osc52').copy('*'),
  },
  paste = {
    ['+'] = require('vim.ui.clipboard.osc52').paste('+'),
    ['*'] = require('vim.ui.clipboard.osc52').paste('*'),
  },
}
if vim.env.TMUX ~= nil then
  local copy = {'tmux', 'load-buffer', '-w', '-'}
  local paste = {'bash', '-c', 'tmux refresh-client -l && sleep 0.05 && tmux save-buffer -'}
  vim.g.clipboard = {
    name = 'tmux',
    copy = {
      ['+'] = copy,
      ['*'] = copy,
    },
    paste = {
      ['+'] = paste,
      ['*'] = paste,
    },
    cache_enabled = 0,
  }
end
```

**Comment by juru1234** [comment]
@andreypopp Thank you!!! Maybe it would be helpful to have something like this in the neovim documentation?

**Comment by andreypopp** [comment]
Maybe, though I think neovim should just do this as a part of tmux clipboard provider maybe?

**Comment by juru1234** [comment]
I'm not really sure maybe some people do not want to have the copied content in their system clipboard....

**Comment by juru1234** [comment]
I think this is not a full solution to the problem. If you ssh from your tmux session to a remote system and run neovim there, this does not work.

**Reply by andreypopp** [reply]
Yes, I usually do ssh then tmux.

**Reply by ntyunyayev** [reply]
I am running into the same issue :/. Did you find a workaround in the end ?

**Reply by juru1234** [reply]
No, I'm pasting with ctrl+shift+v

**Reply by ntyunyayev** [reply]
What about copying? I was thinking about using the tmux copy directly, but it isn't perfect :(

**Reply by juru1234** [reply]
Copying should work flawlessly everywhere.

```
set-option -s set-clipboard on
set -g allow-passthrough on
```

This should be all you need to set in your tmux config. Then, activate OSC52 copy in your neovim config, e.g.

```
-- Use osc52 as clipboard provider
local function paste()
  return {vim.fn.split(vim.fn.getreg(''), '\n'), vim.fn.getregtype('')}
end
vim.g.clipboard = {
	name = 'OSC 52',
	copy = {
		['+'] = require('vim.ui.clipboard.osc52').copy('+'),
		['*'] = require('vim.ui.clipboard.osc52').copy('*'),
	},
	paste = {
		['+'] = paste,
		['*'] = paste,
	},
}
-- To ALWAYS use the clipboard for ALL operations
-- (instead of interacting with the "+" and/or "*" registers explicitly):
vim.opt.clipboard = "unnamedplus"
```

**Reply by ntyunyayev** [reply]
It works, thanks ! In my case, I had "set-option -s set-clipboard external" which was causing issues. I can live with using Ctrl-shift-v to paste, I guess :d

**Reply by ntyunyayev** [reply]
I don't know what changed exactly, but I now use OSC52 for both copying and pasting, and it works even through SSH->TMUX->SSH->NVIM

**Comment by gpanders** [comment]
Neovim will use `tmux load-buffer` and `tmux save-buffer` to read and write the clipboard contents when inside tmux automatically, with no user configuration required. https://github.com/neovim/neovim/blob/5c1b057f8a48c1fbcd2e02fa132ff3ed8dd227a1/runtime/autoload/provider/clipboard.vim#L149-L160

Tmux uses OSC 52 itself to communicate with the "parent" terminal emulator, but it does not support pasting with OSC 52 from applications running inside of tmux (it does not forward the response from the terminal emulator back to the application). So using the OSC 52 plugin from Neovim in tmux will only work for copying.

**Comment by Lcchy** [comment]
FYI here is the related issue tracking this in the tmux repo:
https://github.com/tmux/tmux/issues/4275
